Joss paper (#130)
glotzerlab/coxeter: paper/paper.md

Some examples of such properties are inertia tensors, which are necessary to completely define the equations of motion for anisotropic bodies, and scattering form factors, which are essentially Fourier transforms of the shape volume that are critical to characterizing structure in condensed matter physics.
Since physical equations and observables can be highly sensitive to inputs, coxeter emphasizes robust methods for their evaluation [@Kallay2006].
A number of less commonly used metrics, such as mean curvatures and asphericities, are also provided to support specific research avenues [Irrgang2017].

The library also serves as a repository for the generation of shapes.
While simple classes of shapes like spheres and ellipsoids can be described via a small fixed set of parameters, the definitions of polygons and polyhedra can be arbitrarily long depending on the number of vertices of these shapes.
The shape family API in coxeter provides a flexible way to define and work with collections of related shapes, ranging from enumerable sets like the Platonic solids to continuously defined sets of shapes [@Chen2014].

> These different types of shape families are handled using identical APIs, so users can easily switch between shapes that have completely different mathematical definitions using a single line of code.

Author: I think I missed this one, and given the current stage of the draft I'm not sure what to do with this. Have a read through the current draft then let's see how we feel?

Additionally, since these shape families generate the coxeter's shape classes, calculating the various attributes of shapes is as simple as querying the properties of objects generated from families.

# Statement of Need

Considerations of shape are becoming increasingly important in materials science as improved synthetic capabilities have allowed the creation of a wide range of anisotropic particles [@Glotzer2007b].
Colloidal science in particular has seen immense growth in this area, and numerous studies have shown that particle shape is an important handle for controlling the self-assembly of colloidal crystals [@Damasceno2012d,@Glotzer2007b,@Chen2014].

> Precise modeling of these systems requires robust, reproducible methods for generating shapes and calculating their properties [@Anderson2020,@Ramasubramani2020b,@Allen2006], a gap that coxeter aims to fill.

Author: I think HOOMD and the Allen paper are appropriate citations for this particular sentence, since they are examples of why we need robust calculations of these properties: for the types of dynamics algorithms encoded by tools like this. However, they're also basically two random examples out of hundreds of possibly valid citations, so maybe we should just remove them since it's a pretty obvious statement?

Regarding papers that actually use coxeter/euclid: the citation of my ALJ paper is actually relevant of a project that used this, and since it's the only published work I'm aware of that has used coxeter since its rewrite/conversion from euclid, it's probably the best citation that we have. From a quick search for `import euclid` in our nfs space (only up to very limited depth, so I'm sure I missed scripts that are deeply nested somewhere in the directory structure), projects using euclid or coxeter are:

- All of Bryan's, so I think we can safely cite any of his published papers.
- Rosy's TetragonalDiamond project, both in the photonics code and in PMFT calculations. Since the PMFTs are in her paper, I think we can probably cite it (the Phys Rev Mat one).
- Kelly's dodecagonal quasicrystal project, which has currently been submitted. I don't know what the status is, but it's possible that it'll get published by the time that this paper is ready, in which case we could include it.
- Yein's bipyramid work, which I have no idea what state it's in.
- Sangmin's two step nucleation project, which is published but his usage of euclid seems super minimal so without asking him for more details I wouldn't want to cite it as an example.

The other question is where discussion of papers using coxeter fits best in the new draft.

> Precise modeling of these systems requires robust, reproducible methods for generating shapes and calculating their properties [@Anderson2020,@Ramasubramani2020b,@Allen2006], a gap that coxeter aims to fill.

Author: @bdice any more thoughts on this front?

We anticipate that the set of shape families in coxeter will grow over time as users generate and contribute their shape families to coxeter, with the goal of providing a centralized repository for use in reproducing and extending prior research.

The shape classes in coxeter address the second stated requirement, the ability to perform robust calculations of various properties of shapes.
Most computational geometry libraries typically focus on solving more complex problems like finding convex hulls, Voronoi tesselations, and Delaunay triangulations [@cgal].
The purpose of coxeter is to provide a standard implementation for simpler calculations such as volumes, bounding spheres, and inertia tensors, for which formulas are generally well-known but often require careful consideration to implement robust and efficient algorithms to compute.
Inertia tensors are a particularly notable example since they can be quite difficult to compute accurately for arbitrary polytopes and require specialized algorithms [@Kallay2006].

> Physics-based models, for instance molecular dynamics simulations based on fundamental laws of motion, can be highly sensitive to the precise values of these quantities, necessitating their accurate evaluation.

Author: Rather than a specific implementation section, I've interspersed these citations where I think they're most appropriate. Let me know if you think there's a better way to do this.

# Acknowledgements

> **Update acknowledgments**

Author: Yep I'll update this whole thing at the end. Right now it's just people who I don't want to forget, I'm not worried about funding acknowledgments.

---
title: 'coxeter: A Python package for working with shapes'
tags:
  - Python
  - geometry
  - physics
  - materials science
authors:
  - name: Vyas Ramasubramani
    orcid: 0000-0001-5181-9532
    affiliation: 1
  - name: Bradley Dice
    orcid: 0000-0002-9983-0770
    affiliation: 3
  - name: Tobias Dwyer
    orcid: 0000-0001-6443-7744
    affiliation: 1
  - name: Sharon C. Glotzer
    orcid: 0000-0002-7197-0085
    affiliation: "1, 2, 3, 4"
affiliations:
 - name: Department of Chemical Engineering, University of Michigan
   index: 1
 - name: Department of Materials Science and Engineering, University of Michigan
   index: 2
 - name: Department of Physics, University of Michigan
   index: 3
 - name: Biointerfaces Institute, University of Michigan
   index: 4
date: 8 Dec 2020
bibliography: paper.bib
---

# Package Overview

> The coxeter Python package provides the tools to represent, generate, and compute properties of shapes in two and three dimensions.

Reviewer: ```suggestion
The coxeter Python package provides tools to represent, generate, and compute properties of shapes in two and three dimensions.
```

> The package emphasizes simplicity and flexibility, using a common set of abstractions to present a largely uniform interface across various shapes and allowing easy mutation of almost all of their geometric attributes.

Reviewer: This is an _excellent_ sentence.

Author: Thanks :)

> The package also aims to function as a repository for specific groups of shapes, exposing an easy-to-use API for shape generation that users can easily extend to make particular sets of shapes collectively accessible.

Reviewer: - "aims to function" is too verbose. Replaced with "serves"
- "easily" is redundant with "easy-to-use"
- "specific groups of shapes" is redundant with "particular sets of shapes." I suggested "particular geometric objects" but that may be less precise.
```suggestion
The package also serves as a repository for specific groups of shapes, exposing an easy-to-use API for shape generation that users can extend to make particular geometric objects collectively accessible.
```

Author: First two changes are definitely good. I think the third is an improvement, maybe not what we want in the final version but we can reassess in future drafts.

# Statement of Need

> Considerations of shape are becoming increasingly important in materials science as improved synthetic capabilities have allowed the creation of a wide range of anisotropic particles [@Glotzer2007b].
> Colloidal science in particular has seen immense growth in this area, and numerous studies have shown that particle shape is an important handle for controlling the self-assembly of colloidal crystals [@Damasceno2012d,@Glotzer2007b,@Chen2014].

Reviewer: I think we need to mention something like "colloidal assembly" or "crystals of nanoparticles" earlier because otherwise there's no clear connection between "shape" and "materials science." I want to swap some bits between this sentence and the following sentence. Here's my proposal:
```suggestion
Materials science has recently seen advances in nanoparticle synthesis, allowing the creation of a wide range of anisotropic (non-spherical) particles [@Glotzer2007b].
Consequently, the idea of particle shape in colloidal and nanoscale assembly has seen immense growth.
Numerous studies have shown that particle shape is an important handle for controlling the self-assembly of colloidal crystals [@Damasceno2012d,@Glotzer2007b,@Chen2014].
```

Author: I'm not sure that your changes are addressing your concern. When you say earlier, do you mean closer to the beginning of the SOP? Although in your new version you've shortened the first sentence a bit, the content hasn't really changed, so the same problem persists, unless you really are concerned that the second sentence in the original is too long and so mention of self-assembly at the end is "too late". Perhaps it would be better to swap "anisotropic particles" for "particles with nontrivial shapes" in the first sentence, or something similar to indicate the shape aspect?

Author: @bdice thoughts?

Reviewer: I'm resolving this with no action. It's fine as it is.

Precise modeling of these systems requires reproducible methods for generating shapes and calculating their properties [@Anderson2020,@Allen2006] **Add more citations of real shapes**.

> An important aspect of achieving this reproducibility is making canonical definitions of shapes used in particular studies readibly available to other researchers.

Reviewer: ```suggestion
An important aspect of achieving this reproducibility is making canonical definitions of shapes used in particular studies readily available to other researchers.
```

Furthermore, since these shapes may be used in physics-based simulations, any calculations must be robust enough to handle any numerical issues that may arise across a wide range of different geometries.

# Summary of Features

> ## Shape Analysis **better title?**

Reviewer: Computing Geometric and Physical Properties of Shapes? It's a long subsection title but it summarizes the intent.

Author: If we drop "of Shapes" then that seems good: "Computing Geometric and Physical Properties".

The central elements in coxeter are the shape classes, which encode the features of particular types of shapes.

> In order to enforce a highly uniform API and ensure conceptual clarity, all shape classes extend a small set of abstract base classes that encode specific subsets of properties: for instance, the standard properties of all shapes in two dimensions are embedded in the ``Shape2D`` class.

Reviewer: ```suggestion
In order to enforce a highly uniform API and ensure conceptual clarity, all shape classes inherit from a small set of abstract base classes that encode specific subsets of properties: for instance, the standard properties of all shapes in two dimensions are embedded in the ``Shape2D`` class.
```

> In addition to standard types of shapes such as ellipsoids or polygons, coxeter also includes more esoteric shape classes like spheropolyhedra, which are important to modeling the imperfect rounded polyhedra frequently synthesized at the nano- and micron scales.

Reviewer: ```suggestion
In addition to standard types of shapes such as ellipsoids or polygons, coxeter also includes more esoteric shape classes like spheropolyhedra, which are important for modeling the imperfect rounded polyhedra frequently synthesized at the nano- and micron scales.
```

> Even simple properties like surface areas are generally nontrivial to compute for such shapes, but using coxeter they are no more difficult to work with than other, more common, shapes.

Reviewer: Common is hard to define. Also I lost the antecedent "spheropolyhedra" somewhere. Let's restate it for clarity.
```suggestion
Even simple properties like surface areas are generally nontrivial to compute for such shapes, but using coxeter spheropolyhedra are no more difficult to work with than any other 3D shape.
```

Working with convex polygons and polyhedra using coxeter is greatly simplified via internal use of SciPy's convex hull calculations [@Virtanen2020], allowing the user to simply provide a set of vertices while coxeter performs the appropriate determination of facets and plane equations based on the simplices of the convex hull.

The shape classes transparently expose many geometric attributes in the form of settable Python properties, allowing on-the-fly rescaling or transformation of the shape.
This aspect of coxeter is designed to fill a common gap in most computational geometry libraries, which typically focus on solving more complex problems like finding convex hulls, Voronoi tessellations, and Delaunay triangulations [@cgal]; coxeter aims to provide a standard implementation for simpler calculations such as surface areas and bounding spheres for which formulas are generally well-known but often require careful consideration to calculate robustly and efficiently.

> These properties range from standard calculations like volumes and surface areas to less common metrics like mean curvatures and asphericities that are relevant for specific research avenues [@Irrgang2017].

Reviewer: I'd insert a concrete phrase instead of "specific research avenues," it sounds really esoteric and doesn't indicate the scientific utility of this package. e.g. "Understanding how shape properties correspond to self-assembly behavior of colloidal crystals."

Author: How about "computing equations of state for nonspherical particles"?

The package also provides various types of bounding and bounded spheres of shapes, which are critical measures of the extent of polygons and polyhedra within crystal structures.
To simplify interoperability with other packages in the scientific computing ecosystem, non-scalar properties are generally provided as NumPy arrays [@Harris2020].

In addition to purely geometric properties, shapes in coxeter also expose various physically relevant quantities in order to support a wide range of applications.
Some examples of such properties are inertia tensors, which are integral to the equations of motion for anisotropic bodies, and scattering form factors, which are essentially Fourier transforms of the shape volume that are critical to characterizing structure in condensed matter physics.
Since physical equations and observables can be highly sensitive to inputs like inertia tensors, coxeter emphasizes robust methods for their evaluation [@Kallay2006].

> Two dimensional shapes like polygons are embedded in three dimensions rather than in the plane, so coxeter uses the rowan library [@Ramasubramani2018] to rotate them into the plane and then compute various properties to avoid complications and numerical instabilitiesthat arise from performing integrals over planar lamina embedded in $\mathcal{R}^3$.

Reviewer: If you disagree with my second edit and wish to use R^3 instead of "3D Euclidean space," I think we want `\mathbb{R}^3` over `\mathcal{R}^3`.
```suggestion
Two dimensional shapes like polygons are embedded in three dimensions rather than in the plane, so coxeter uses the rowan library [@Ramasubramani2018] to rotate them into the plane and then compute various properties to avoid complications and numerical instabilities that arise from performing integrals over planar lamina embedded in 3D Euclidean space.
```

## Shape Generation

The library also serves as a repository for the generation of shapes.
While simple classes of shapes like spheres and ellipsoids can be described via a small fixed set of parameters, the definitions of polygons and polyhedra can be arbitrarily long depending on the number of vertices of these shapes.
The shape family API in coxeter provides a flexible way to define and work with collections of related shapes, ranging from enumerable sets like the Platonic solids to continuously defined sets of shapes [@Chen2014].
These different types of shape families are handled using identical APIs, so users can easily switch between shapes that have completely different mathematical definitions using a single line of code.
Shape families generate coxeter shape classes from input parameters, simplifying access to computed geometric and physical properties.

> A number of such families are bundled into coxeter, but just as importantly, the framework allows users to work with arbitrary lists of shapes provided as dictionaries of attributes corresponding to specified schemata.

Reviewer: ...schemata? Too strange of a word. I don't even know what the end of the sentence means, so I deleted it.
```suggestion
A number of such families are bundled into coxeter, but just as importantly, the framework allows users to work with arbitrary lists of shapes provided as dictionaries of attributes.
```

This dictionary-based definition can be trivially converted to JSON, making it trivial to share representations of shapes.
The library also stores mappings from digital object identifiers (DOIs) to families, so that any user can contribute families associated with published research to make them immediately collectively accessible.
We anticipate that the set of shape families in coxeter will grow over time as users generate and contribute their shape families to coxeter, with the goal of providing a centralized repository for use in reproducing and extending prior research, particularly in the field of shape-driven nanoparticle self-assembly.

> Currently coxeter primarily supports the schema proposed by the GSD library **should we cite? there's no canonical source other than RTD**, making it directly compatible with the HOOMD-blue molecular simulation tool [@Anderson2020], but other schema can easily be implemented on an as-needed basis.

Reviewer: Yes, we should cite GSD as if it were a web page, and point to the GitHub source.

---
title: 'coxeter: A Python package for working with shapes'
tags:
  - Python
  - geometry
  - physics
  - materials science

> authors:
>   - name: Vyas Ramasubramani
>     orcid: 0000-0001-5181-9532
>     affiliation: 1
>   - name: Bradley Dice
>     orcid: 0000-0002-9983-0770
>     affiliation: 3
>   - name: Tobias Dwyer
>     orcid: 0000-0001-6443-7744
>     affiliation: 1
>   - name: Sharon C. Glotzer
>     orcid: 0000-0002-7197-0085
>     affiliation: "1, 2, 3, 4"
> affiliations:
>  - name: Department of Chemical Engineering, University of Michigan
>    index: 1
>  - name: Department of Materials Science and Engineering, University of Michigan
>    index: 2
>  - name: Department of Physics, University of Michigan
>    index: 3
>  - name: Biointerfaces Institute, University of Michigan
>    index: 4

Reviewer: This needs to be re-ordered so that the numbers are "Ramasubramani (1), Dice (2), Dwyer (1), Glotzer (1, 2, 3, 4)."
```suggestion
authors:
  - name: Vyas Ramasubramani
    orcid: 0000-0001-5181-9532
    affiliation: 1
  - name: Bradley Dice
    orcid: 0000-0002-9983-0770
    affiliation: 2
  - name: Tobias Dwyer
    orcid: 0000-0001-6443-7744
    affiliation: 1
  - name: Sharon C. Glotzer
    orcid: 0000-0002-7197-0085
    affiliation: "1, 2, 3, 4"
affiliations:
 - name: Department of Chemical Engineering, University of Michigan
   index: 1
 - name: Department of Physics, University of Michigan
   index: 2
 - name: Department of Materials Science and Engineering, University of Michigan
   index: 3
 - name: Biointerfaces Institute, University of Michigan
   index: 4
```

These different types of shape families are handled using identical APIs, so users can easily switch between shapes that have completely different mathematical definitions using a single line of code.
Shape families generate coxeter shape classes from input parameters, simplifying access to computed geometric and physical properties.

A number of such families are bundled into coxeter, but just as importantly, the framework allows users to work with arbitrary lists of shapes provided as dictionaries of attributes.
This dictionary-based definition can be simply converted to JSON, making it trivial to share representations of shapes.
The library also stores mappings from digital object identifiers (DOIs) to families, so that any user can contribute families associated with published research to make them immediately collectively accessible.
We anticipate that the set of shape families in coxeter will grow over time as users generate and contribute their shape families to coxeter, with the goal of providing a centralized repository for use in reproducing and extending prior research, particularly in the field of shape-driven nanoparticle self-assembly.

> Currently coxeter primarily supports the schema proposed by the GSD library [@GlotzerLabGSD], making it directly compatible with the HOOMD-blue molecular simulation tool [@Anderson2020], but other schema can easily be implemented on an as-needed basis.

Reviewer: ```suggestion
Currently coxeter primarily supports the schema proposed by the GSD library [@GlotzerLabGSD], making it directly compatible with the HOOMD-blue molecular simulation tool [@Anderson2020], but other schema can be implemented as needed.
```
